node/core/parachains-inherent/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The parachain inherent data provider
+//!
+//! Parachain backing and approval is an off-chain process, but the parachain needs to progress on chain as well. To
+//! make it progress on chain a block producer needs to forward information about the state of a parachain to the
+//! runtime. These information are forwarded through an inherent to the runtime. Here we provide the
+//! [`ParachainInherentDataProvider`] that requests the relevant data from the provisioner subsystem and creates the
+//! the inherent data that the runtime will use to create an inherent.
+
+#![deny(unused_crate_dependencies, unused_results)]
+
+use futures::{select, FutureExt};
+use polkadot_node_subsystem::{
+	messages::{AllMessages, ProvisionerMessage}, SubsystemError,
+};
+use polkadot_overseer::OverseerHandler;
+use polkadot_primitives::v1::{
+	Block, Hash, InherentData as ParachainsInherentData,
+};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::generic::BlockId;
+use std::time;
+
+/// How long to wait for the provisioner, before giving up.
+const PROVISIONER_TIMEOUT: time::Duration = core::time::Duration::from_millis(2500);
+
+/// Provides the parachains inherent data.
+pub struct ParachainsInherentDataProvider {
+	inherent_data: ParachainsInherentData,
+}
+
+impl ParachainsInherentDataProvider {
+	/// Create a new instance of the [`ParachainsInherentDataProvider`].
+	pub async fn create<C: HeaderBackend<Block>>(
+		client: &C,
+		mut overseer: OverseerHandler,
+		parent: Hash,
+	) -> Result<Self, Error> {
+		let pid = async {
+			let (sender, receiver) = futures::channel::oneshot::channel();
+			overseer.wait_for_activation(parent, sender).await;
+			receiver.await.map_err(|_| Error::ClosedChannelAwaitingActivation)?.map_err(Error::Subsystem)?;
+
+			let (sender, receiver) = futures::channel::oneshot::channel();
+			overseer.send_msg(AllMessages::Provisioner(
+				ProvisionerMessage::RequestInherentData(parent, sender),
+			)).await;
+
+			receiver.await.map_err(|_| Error::ClosedChannelAwaitingInherentData)
+		};
+
+		let mut timeout = futures_timer::Delay::new(PROVISIONER_TIMEOUT).fuse();
+
+		let parent_header = match client.header(BlockId::Hash(parent)) {
```

Comment:
can we use a chain api message instead of passing client directly? we're trying to avoid client creep into every subsystem

Reply:
This is no subsystem.

Reply:
previously we passed client into proposer because we had to, now we don't, but yeah, maybe we actually want higher priority than other chain api requestors